### PR TITLE
[Feature]Add transformer in transformer

### PR DESCRIPTION
**Pull request by HIT-cwh**
Hi! This pr currently provided the following features:

1. The model of TnT, including the TnT backbone, TnT classification head, etc
2. Some configs for TnT inference
3. This model achieves the same accuracy with the pretrained weights from pytorch-image-models(https://github.com/contrastive/pytorch-image-models/releases/download/TNT/tnt_s_patch16_224.pth.tar)

Pls take a look.

### CODE REVIEW

**Comment by mzr1996** [comment]
Please modify the config file name, which refers to #355.
And please add basic optimizer config, at least, users shouldn't get error if they run
```shell
python tools/train.py configs/tnt/tnt_s_patch16_224_inference.py
```

File: mmcls/models/backbones/tnt.py [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+                                      self.new_patch_size[1])
+        x = x + pixel_pos
+        x = x.reshape(B * self.num_patches, self.inner_dims,
+                      -1).transpose(1, 2)
+        return x
+
+
+@BACKBONES.register_module()
+class TNT(BaseBackbone):
+    """ Transformer in Transformer - https://arxiv.org/abs/2103.00112
```
Comment (mzr1996) [inline mmcls/models/backbones/tnt.py:192]: Please format the docstring, and add the reference source.
Refers to https://github.com/open-mmlab/mmclassification/blob/1cab2e5b634c073998a9bd093b8baa6cf18fe2ce/mmcls/models/backbones/swin_transformer.py#L176-L182

**Review by mzr1996 — CHANGES_REQUESTED** [review]

File: tests/test_classifiers.py [hunk 2]
```diff
@@ -219,3 +219,45 @@ def test_image_classifier_vit():
 
     losses = img_classifier.forward_train(imgs, label)
     assert losses['loss'].item() > 0
+
+
+def test_image_classifier_tnt():
```
Comment (mzr1996) [inline tests/test_classifiers.py:224]: Remove this unit test. Place all backbone unit tests in `tests/test_models/test_backbones`

File: tests/test_backbones/test_tnt.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+def check_norm_state(modules, train_state):
+    """Check if norm layer is in correct train state."""
+    for mod in modules:
+        if isinstance(mod, _BatchNorm):
+            if mod.training != train_state:
+                return False
+    return True
+
+
+def test_tnt_backbone():
```
Comment (mzr1996) [inline tests/test_backbones/test_tnt.py:17]: Place all backbone unit tests in `tests/test_models/test_backbones`. Refers to #321

File: mmcls/models/backbones/tnt.py [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+            self.norm_proj(pixel_embed).reshape(B, N - 1, -1))
+        patch_embed = self.attn_out(
+            self.norm_out(patch_embed), identity=patch_embed)
+        patch_embed = self.ffn_out(
+            self.norm_ffn_out(patch_embed), identity=patch_embed)
+
+        return pixel_embed, patch_embed
+
+
+class PixelEmbed(BaseModule):
```
Comment (mzr1996) [inline mmcls/models/backbones/tnt.py:127]: Move to `mmcls/models/utils/embed.py` if it's universal.

File: mmcls/models/backbones/tnt.py [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+from ..utils import to_2tuple
+from .base_backbone import BaseBackbone
+
+
+class TnTLayer(BaseModule):
+    """Implement one encoder layer in Transformer in Transformer.
+
+    Args:
+        inner_dims (int): The feature dimension in inner transformer block
+        embed_dims (int): The feature dimension in outer transformer block
```
Comment (mzr1996) [inline mmcls/models/backbones/tnt.py:20]: `inner_dims` and `embed_dims` are a little weird.
`inner_embed_dims` and `outer_embed_dims` may be better.

File: mmcls/models/backbones/tnt.py [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+                 qkv_bias=False,
+                 act_cfg=dict(type='GELU'),
+                 norm_cfg=dict(type='LN'),
+                 batch_first=True,
+                 init_cfg=None):
+        super(TnTLayer, self).__init__(init_cfg=init_cfg)
+
+        # Inner transformer
+        self.norm_in = build_norm_layer(norm_cfg, inner_dims)[1]
+        self.attn_in = MultiheadAttention(
```
Comment (mzr1996) [inline mmcls/models/backbones/tnt.py:64]: Use `inner` and `outer` instead of `in` and `out`.

File: mmcls/models/backbones/tnt.py [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+        # outer
+        B, N, C = patch_embed.size()
+        patch_embed[:, 1:] = patch_embed[:, 1:] + self.projection(
+            self.norm_proj(pixel_embed).reshape(B, N - 1, -1))
+        patch_embed = self.attn_out(
+            self.norm_out(patch_embed), identity=patch_embed)
+        patch_embed = self.ffn_out(
+            self.norm_ffn_out(patch_embed), identity=patch_embed)
+
+        return pixel_embed, patch_embed
```
Comment (mzr1996) [inline mmcls/models/backbones/tnt.py:124]: If the "inner transformer block" and "outer transformer block" structures are almost the same, can we encapsulate them as one module?
```suggestion
    def forward(self, pixel_embed, patch_embed):
        pixel_embed = self.inner_block(pixel_embed)

        B, N, C = patch_embed.size()
        patch_embed[:, 1:] = patch_embed[:, 1:] + self.projection(
            self.norm_proj(pixel_embed).reshape(B, N - 1, -1))
            
        patch_embed = self.outer_block(patch_embed)

        return pixel_embed, patch_embed
```

**Review by mzr1996 — CHANGES_REQUESTED** [review]

File: mmcls/models/backbones/tnt.py (hunk 4 above)
Comment (HIT-cwh) [reply mmcls/models/backbones/tnt.py:138]: To the best of my knowledge, PixelEmbed is only used in TnT .

**Review by Ezra-Yu — APPROVED** [review]
LGTM

**Review by mzr1996 — APPROVED** [review]
LGTM
